### PR TITLE
patch sequence page

### PR DESCRIPTION
**Pull request by c-cube**

### CODE REVIEW

**Comment by cuihtlauac** [comment]
Thanks @c-cube, am I correct assuming this only about the Sequences documentation and the other files went in this PR accidentally?

**Comment by sabine** [comment]
This might be the state of staging at the time.

**Comment by c-cube** [comment]
Ok I have no idea what's going on there. I edited from the "contribute" button on the website (URL: https://github.com/ocaml/ocaml.org/blob/c461890f0ce6473bd4faf55b1d7157c1000e0fcb/data/tutorials/ds_05_seq.md) and ended up on a detached commit?

**Comment by sabine** [comment]
Sorry about that, the contribute link goes to the commit on the `staging` branch.

**Comment by c-cube** [comment]
if someone wants to cherry pick the small change… :)

**Comment by sabine** [comment]
So the diff with `staging` instead of main is going to be much better. 🙂

No worries, we will take it from here.

**Comment by c-cube** [comment]
Thanks y'all :)

File: data/tutorials/ds_05_seq.md [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+type 'a list =
+  | []
+  | (::) of 'a * 'a list
+```
+and `Seq.t`, which is merely a type alias for `unit -> 'a Seq.node`. The whole
+point of this definition is the second argument's type `Seq.Cons`, which is a
+function returning a sequence while its `list` counterpart returns a list. Let's
+compare the constructors of `list` and `Seq.node`:
+1. Empty lists and sequences are defined the same way, a constructor without any
+   parameter: `Seq.Nil` and `[]`.
```
Comment (christinerose) [inline data/tutorials/ds_05_seq.md:43]:
```suggestion
   parameters: `Seq.Nil` and `[]`.
```

File: data/tutorials/ds_05_seq.md [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+reaching `max_int`, the values will circle down to `min_int`. They are
+ultimately periodic.
+
+The OCaml standard library contains a module on sequences called
+[`Seq`](/releases/5.0/api/Seq.html). It contains a `Seq.iter` function, which
+has the same behaviour as `List.iter`. Writing this:
+```ocaml
+# Seq.iter print_int (ints 0);;
+```
+in an OCaml top-level means “print integers forever,” and you have to press
```
Comment (christinerose) [inline data/tutorials/ds_05_seq.md:83]:
```suggestion
in an OCaml toplevel means “print integers forever,” and you have to press
```

File: data/tutorials/ds_05_seq.md [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+ 239; 241; 251; 257; 263; 269; 271; 277; 281; 283; 293; 307; 311; 313; 317;
+ 331; 337; 347; 349; 353; 359; 367; 373; 379; 383; 389; 397; 401; 409; 419;
+ 421; 431; 433; 439; 443; 449; 457; 461; 463; 467; 479; 487; 491; 499; 503;
+ 509; 521; 523]
+```
+
+The function `sieve` is recursive in OCaml and common sense. It is defined using
+the `rec` keyword and calls itself. However, some call that kind of function
+[_corecursive_](https://en.wikipedia.org/wiki/Corecursion). This word is used to
+emphasize that, by design, it does not terminate. Strictly speaking, the sieve
```
Comment (christinerose) [inline data/tutorials/ds_05_seq.md:159]:
```suggestion
emphasise that, by design, it does not terminate. Strictly speaking, the sieve
```

File: data/tutorials/ds_05_seq.md [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+Sequences are functions, as stated by their type:
+```ocaml
+# #show Seq.t;;
+type 'a t = unit -> 'a Seq.node
+```
+Functions working with sequences must be written accordingly.
+* Sequence consumer: partially applied function parameter
+* Sequence producer: partially applied function result
+
+When code dealing with sequences does not behave as expected like if it is
```
Comment (christinerose) [inline data/tutorials/ds_05_seq.md:267]:
```suggestion
When code dealing with sequences does not behave as expected, like if it is
```

File: data/tutorials/ds_05_seq.md [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+  ```ocaml
+  val Array.of_seq : 'a array -> 'a Seq.t
+  val Array.to_seq : 'a Seq.t -> 'a array
+  ```
+* Strings
+  ```ocaml
+  val String.of_seq : string -> char Seq.t
+  val String.to_seq : char Seq.t -> string
+  ```
+Similar functions are also provided for sets, maps, hash tables (`Hashtbl`) and
```
Comment (christinerose) [inline data/tutorials/ds_05_seq.md:291]:
```suggestion
Similar functions are also provided for sets, maps, hash tables, (`Hashtbl`) and
```

File: data/tutorials/gs_00_up_and_running.md [hunk 6]
```diff
@@ -13,8 +13,9 @@ date: 2021-05-27T21:07:30-00:00
...
 
-The following instructions work on Linux, BSD, and macOS. Plus, they also work
-on Cygwin and WSL. If you want to set up OCaml on Windows, you might be
+We provide installation instructions for Linux, macOS, and *BSD for all OCaml versions. For Windows, we provide instructions on this page for installing OCaml 4.14.0 via the [Diskuv OCaml](https://github.com/diskuv/dkml-installer-ocaml#readme) Installer. Note that, if you use Windows Subsystem for Linux (WSL), the Unix instructions can be used on Windows.
+
+If you are setting up OCaml on Windows and are unsure which installation method to use, you might be
 interested in reading [OCaml on Windows](/docs/ocaml-on-windows) first.
 
 **Guidelines for Following Instructions on this Page**
```
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md:21]:
```suggestion
**Guidelines for Following Instructions on This Page**
```

File: data/tutorials/gs_00_up_and_running.md [hunk 7]
```diff
@@ -28,16 +29,12 @@ The code blocks (in black) on this page show the required commands (the text aft
...
 
-OCaml is available as a package in most Linux distributions; however, it is
-often outdated. On the contrary, OCaml's official package manager, opam, allows you to
-easily switch between OCaml versions and much more. This is
-especially useful since different projects might require different versions of
-OCaml.
-
-On Unix, the best way to install OCaml is with opam, OCaml's package manager. On Windows, the best way is to use a traditional `setup.exe` that will initialise opam
-on your behalf; however, only OCaml version 4.14.0 can be installed with `setup.exe` now.
+OCaml has an official package manager, `opam`, which allows you to
```
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md:32]:
```suggestion
OCaml has an official package manager, opam, which allows you to
```

File: data/tutorials/gs_00_up_and_running.md [hunk 8]
```diff
@@ -28,16 +29,12 @@ The code blocks (in black) on this page show the required commands (the text aft
...
+easily switch between OCaml versions and and much more.
```
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md:33]:
```suggestion
easily switch between OCaml versions and much more.
```

File: data/tutorials/gs_00_up_and_running.md [hunk 9]
```diff
@@ -28,16 +29,12 @@ The code blocks (in black) on this page show the required commands (the text aft
...
+It enables you to conveniently deal with different projects which
```
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md:34]:
```suggestion
It enables you to conveniently deal with different projects that
```

File: data/tutorials/gs_00_up_and_running.md [hunk 10]
```diff
@@ -28,16 +29,12 @@ The code blocks (in black) on this page show the required commands (the text aft
...
+require different versions of OCaml.
 
-[opam](https://opam.ocaml.org/), OCaml's package manager, introduces the concept of "switches," which is a compiler with a set of packages (libraries and other files). Switches are used to have independent sets of dependencies in different projects.
+[`opam`](https://opam.ocaml.org/) introduces the concept of a "switch," which is an isolated environment that contains an OCaml compiler together with a set of packages (libraries and other files). Switches allow us to install independent sets of dependencies for different projects.
```
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md:37]:
```suggestion
[opam](https://opam.ocaml.org/) introduces the concept of a "switch," which is an isolated environment that contains an OCaml compiler together with a set of packages (libraries and other files). Switches allow us to install independent sets of dependencies for different projects.
```

File: data/tutorials/gs_00_up_and_running.md [hunk 11]
```diff
@@ -53,9 +50,15 @@ below](#up-and-running-with-the-platform-installer) for the instructions.
 However, please note that it is still experimental and in active development
 (any feedback is highly appreciated).
 
-### Installation for Unix
+### Installation on Unix (Linux, macOS, *BSD)
+
+Note: OCaml is available as a package in most Linux distributions; however, it is
+often outdated. The best way to install OCaml is with `opam`, OCaml's package manager.
```
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md:56]:
```suggestion
often outdated. The best way to install OCaml is with opam, OCaml's package manager.
```

File: data/tutorials/gs_00_up_and_running.md [hunk 12]
```diff
@@ -53,9 +50,15 @@ below](#up-and-running-with-the-platform-installer) for the instructions.
...
 
-After having installed opam, you will need to initialise it, [see below](#initialising-opam-on-unix).
+The following steps depend on having these packages or tools installed:
+`gcc`, `build-essential`, `curl`, `bubblewrap`, and `unzip`.
+
+#### 1. Install opam
 
 To install opam, you can [use your system package manager](https://opam.ocaml.org/doc/Install.html#Using-your-distribution-39-s-package-system) or download the [binary distribution](https://opam.ocaml.org/doc/Install.html#Binary-distribution). The details are available in the above links, but for convenience, we copy a few of them here:
```
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md:63]:
```suggestion
To install opam, you can [use your system package manager](https://opam.ocaml.org/doc/Install.html#Using-your-distribution-39-s-package-system) or download the [binary distribution](https://opam.ocaml.org/doc/Install.html#Binary-distribution). The details are available in the above links, but for convenience, we copied a few of them here:
```

File: data/tutorials/gs_00_up_and_running.md [hunk 13]
```diff
@@ -123,9 +122,21 @@ The second command (`eval $(opam env)`) modifies a few environment variables to
 
 Now check the installation by running `opam --version`. You can compare it with the current version on [opam.ocaml.org](https://opam.ocaml.org/).
 
-#### Creating a New Switch on Unix
+**Please note:** In case you are running `opam init` inside a Docker container,
+you will be asked whether you want to disable sanboxing.
```
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md:126]:
```suggestion
you will be asked whether you want to disable sandboxing.
```

File: data/tutorials/gs_00_up_and_running.md [hunk 14]
```diff
@@ -123,9 +122,21 @@ The second command (`eval $(opam env)`) modifies a few environment variables to
...
+This is necessary, unless you run a privileged Docker container.
 
-If you want a specific version of OCaml, or a new independent environment, you can create a new switch with the `opam switch create` command. Specify which version as shown below (i.e., `opam switch create 4.14.0`). All possible compiler versions can be found with `opam switch list-available`. The most current version can be found at [opam.ocaml.org](https://opam.ocaml.org/packages/ocaml-base-compiler/).
+#### 3. Create an `opam` Switch
```
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md:129]:
```suggestion
#### 3. Create an opam Switch
```

File: data/tutorials/gs_00_up_and_running.md [hunk 15]
```diff
@@ -123,9 +122,21 @@ The second command (`eval $(opam env)`) modifies a few environment variables to
...
+
+This step is necessary only if you need to install a specific version of OCaml,
+or if you want to create a new independent environment.
+`opam init` already set up a default `opam` switch for you to work in.
```
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md:133]:
```suggestion
`opam init` already set up a default opam switch for you to work in.
```

File: data/tutorials/gs_00_up_and_running.md [hunk 16]
```diff
@@ -146,17 +157,23 @@ We will learn about the OCaml toplevel and other installed tools in the next sec
...
+
+Note that only OCaml version 4.14.0 is available via Diskuv OCaml.
 
 > Advanced Users: If you are familiar with Cygwin or WSL2, there are other installation methods described on the [OCaml on Windows](/docs/ocaml-on-windows) page.
 
+#### 1. Use the DKML Installer
+
 Before using the DKML installer, briefly review the following:
 
+* Do not use the installer if you have a space in your username (ex. `C:\Users\Jane Smith`).
```
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md:172]:
```suggestion
* Do not use the installer if you have a space in your username (e.g., `C:\Users\Jane Smith`).
```

File: data/tutorials/gs_00_up_and_running.md [hunk 17]
```diff
@@ -178,11 +195,15 @@ After the DKML installer finishes installing two programs (`Visual Studio Instal
 
 Now, download and run:
 
-* OCaml 4.14.0 with Git and Visual Studio compiler: [setup-diskuv-ocaml-windows_x86_64-1.1.0.exe](https://github.com/diskuv/dkml-installer-ocaml/releases/download/v1.1.0_r2/setup-diskuv-ocaml-windows_x86_64-1.1.0.exe)
+* OCaml 4.14.0 with Git and Visual Studio compiler: [setup-diskuv-ocaml-windows_x86_64-1.2.0.exe](https://github.com/diskuv/dkml-installer-ocaml/releases/download/v1.2.0/setup-diskuv-ocaml-windows_x86_64-1.2.0.exe)
 
-#### Creating a New Switch on Windows
+#### 2. Create an `opam` Switch
```
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md:200]:
```suggestion
#### 2. Create an opam Switch
```

File: data/tutorials/gs_00_up_and_running.md [hunk 18]
```diff
@@ -211,9 +232,10 @@ The OCaml toplevel, version 4.14.0
 To learn more about Diskuv OCaml, see the [official
 Diskuv OCaml documentation](https://diskuv-ocaml.gitlab.io/distributions/dkml/#introduction).
 
-### The OCaml Base Tools
+### The OCaml Base Tools are Now Installed
```
Comment (christinerose) [inline data/tutorials/gs_00_up_and_running.md:235]:
```suggestion
### The OCaml Base Tools Are Now Installed
```

File: data/tutorials/gs_01_a_first_hour_with_ocaml.md [hunk 19]
```diff
@@ -1,50 +1,62 @@
...
 as described in [Up and Running](/docs/up-and-running).
 
 Alternatively, you may follow almost all of it by running OCaml in your browser
 using the [OCaml Playground](https://ocaml.org/play), with no installation required.
 
-On macOS/iOS/iPadOS, you can download this [all-in-one package on the App Store](https://apps.apple.com/app/ocaml-learn-code/id1547506826). It contains an editor side-by-side with an interactive toplevel. Plus, it's free and [open source](https://github.com/GroupeMINASTE/OCaml-iOS)!
+On macOS/iOS/iPadOS, you can download this [all-in-one package on the App Store](https://apps.apple.com/app/ocaml-learn-code/id1547506826). It contains an editor side-by-side with an interactive toplevel. Plus, it's free and [open source](https://github.com/GroupeMINASTE/OCaml-iOS).
 
+**Note:** The OCaml code examples in this tutorial start with a prompt `#`.
+This prompt is not part of the code, it's a hint to you that the code
```
Comment (christinerose) [inline data/tutorials/gs_01_a_first_hour_with_ocaml.md:24]:
```suggestion
This prompt is not part of the code; it's a hint to you that the code
```

File: data/tutorials/gs_01_a_first_hour_with_ocaml.md [hunk 20]
```diff
@@ -1,50 +1,62 @@
...
+following it should be either entered into an interactive toplevel, or
```
Comment (christinerose) [inline data/tutorials/gs_01_a_first_hour_with_ocaml.md:25]:
```suggestion
following it should be either entered into an interactive toplevel or
```

File: data/tutorials/gs_01_a_first_hour_with_ocaml.md [hunk 21]
```diff
@@ -1,50 +1,62 @@
...
+into the OCaml Playground.
 ## Running OCaml Programs
 
-To try small OCaml expressions, use an interactive toplevel, or REPL
-(Read-Eval-Print Loop). The `ocaml` command provides a basic toplevel
-(install `rlwrap` through your system package manager and run `rlwrap ocaml` instead to get history navigation).
+### Using an OCaml toplevel (REPL)
```
Comment (christinerose) [inline data/tutorials/gs_01_a_first_hour_with_ocaml.md:29]:
```suggestion
### Using an OCaml Toplevel (REPL)
```

File: data/tutorials/gs_01_a_first_hour_with_ocaml.md [hunk 22]
```diff
@@ -1,50 +1,62 @@
...
 
-The alternative REPL [UTop](https://github.com/ocaml-community/utop) may be installed
-through [opam](https://opam.ocaml.org/doc/Install.html) or your system package manager. It has
-the same basic interface but is much more convenient to use (history
-navigation, auto-completion, etc.). 
+We recommend using UTop (see [here](/docs/up-and-running#using-the-ocaml-toplevel-with-utop)),
+but alternatively, the basic toplevel of OCaml (`ocaml` command) can also be used. For installation instructions,
```
Comment (christinerose) [inline data/tutorials/gs_01_a_first_hour_with_ocaml.md:32]:
```suggestion
but alternatively, the basic OCaml toplevel (`ocaml` command) can also be used. For installation instructions,
```

File: data/tutorials/gs_01_a_first_hour_with_ocaml.md [hunk 23]
```diff
@@ -70,32 +82,91 @@ $ ocaml
...
+
+### Using the OCaml Playground
+
+The in-browser [OCaml Playground](https://ocaml.org/play), which you can
+use without installing anything, has this interface:
+
+![OCaml Playground](/media/tutorials/ocaml-playground.png)
 
-## Expressions
+There are two windows: the editor window (on the left) is where you write code,
```
Comment (christinerose) [inline data/tutorials/gs_01_a_first_hour_with_ocaml.md:97]:
```suggestion
There are two windows: the editor window (on the left) is where you write code, and
```

File: data/tutorials/gs_01_a_first_hour_with_ocaml.md [hunk 24]
```diff
@@ -104,41 +175,68 @@ Of course, we can have multiple names:
 - : int = 3
 ```
 
-Note that this is still just one expression. We can define a function to do the
-job for any number:
+This defines two variables `a` and `b` with the values of `1` and `2` respectively,
```
Comment (christinerose) [inline data/tutorials/gs_01_a_first_hour_with_ocaml.md:178]:
```suggestion
This defines two variables `a` and `b` with the values of `1` and `2`, respectively,
```

File: data/tutorials/gs_01_a_first_hour_with_ocaml.md [hunk 25]
```diff
@@ -104,41 +175,68 @@ Of course, we can have multiple names:
...
+This defines a function named `square` which has one argument, namely `x`, and
+its result is equal to (`=`) the result of the expression `x * x`. The expression
+that defines the result of a function is called *function body*.
+
+When we apply the function as `square 50;;`, this evaluates to the function body
+of `square` where `x` is bound to `50`, and thus, we get a result of `50 * 50 = 2500`.
+
+When using `let` to define a function, the first identifier is the function name (`square`, above),
+then any additional identifiers are the different arguments of the function. In our example above,
+the `square` function has only one argument `x`.
```
Comment (christinerose) [inline data/tutorials/gs_01_a_first_hour_with_ocaml.md:204]:
```suggestion
When using `let` to define a function, the first identifier is the function name (`square`, above).
Any additional identifiers are the different arguments of the function. In our example above,
the `square` function has only one argument `x`.
```

File: data/tutorials/gs_01_a_first_hour_with_ocaml.md [hunk 26]
```diff
@@ -379,6 +477,16 @@ combines two lists:
 - : int list = [1; 2; 3]
 ```
 
+Non-empty lists have a head (its first element) and a tail (the list of the rest
+of its elements):
```
Comment (christinerose) [inline data/tutorials/gs_01_a_first_hour_with_ocaml.md:481]:
```suggestion
Non-empty lists have a head (its first element) and a tail (the list of its remaining elements):
```

File: data/tutorials/gs_01_a_first_hour_with_ocaml.md [hunk 27]
```diff
@@ -475,8 +586,8 @@ course. Here are some examples of using `map`:
 (The syntax `fun` ... `->` ... is used to build a function without a name, which
 we'll only use in one place in the program.)
 
-We need not give a function all its arguments at once. This is called partial
-application. For example:
+We can apply a function by providing only some of its arguments. This is called *partial
+application* and the value we get back from such a partial application is a new function. For example:
```
Comment (christinerose) [inline data/tutorials/gs_01_a_first_hour_with_ocaml.md:590]:
```suggestion
application*, and the value we get back from such a partial application is a new function. For example:
```

File: data/tutorials/gs_01_a_first_hour_with_ocaml.md [hunk 28]
```diff
@@ -508,16 +619,20 @@ of lists:
 ## Other Built-In Types
 
 We have seen basic data types like `int`, and our first compound data type, the
```
Comment (christinerose) [inline data/tutorials/gs_01_a_first_hour_with_ocaml.md:621]:
```suggestion
We have seen basic data types like `int` and our first compound data type, the
```

File: data/tutorials/gs_01_a_first_hour_with_ocaml.md [hunk 29]
```diff
@@ -508,16 +619,20 @@ of lists:
...
 
 ```ocaml
 # let t = (1, "one", '1');;
 val t : int * string * char = (1, "one", '1')
 ```
 
-Notice how the type is written. Records are like tuples, but they have named
-elements:
+Notice how the type is written: the individual types that make up a tuple
+are separated by `*`.
```
Comment (christinerose) [inline data/tutorials/gs_01_a_first_hour_with_ocaml.md:632]:
```suggestion
Notice how the type is written: `*` separates the individual types that make up a tuple.
```

File: data/tutorials/lg_07_objects.md [hunk 30]
```diff
@@ -375,25 +375,31 @@ Notes:
...
 change attempted by the outside code. This means, for example, that you
 could change the internal representation to use an array at some later
 date, and no code outside the class would need to be changed.
 
 Last, but not least, we have implemented the previously virtual
 `repaint` function so that `container#repaint` will repaint all of the
-contained widgets. Notice I use `List.iter` to iterate over the list,
-and I also use a probably unfamiliar anonymous function expression:
+contained widgets. Notice the use of `List.iter` to iterate over the list,
+and I also use an anonymous function expression that might be unfamiliar:
```
Comment (christinerose) [inline data/tutorials/lg_07_objects.md:393]:
```suggestion
and also the use of an anonymous function expression that might be unfamiliar:
```

File: data/tutorials/lg_07_objects.md [hunk 31]
```diff
@@ -375,25 +375,31 @@ Notes:
...
 # (fun w -> w#repaint);;
 - : < repaint : 'a; .. > -> 'a = <fun>
 ```
-which defines an anonymous function with one argument `w` that just
+This defines an anonymous function with one argument `w` that just
 calls `w#repaint` (the `repaint` method on widget `w`).
 
-In this instance our `button` class is simple (rather unrealistically
+In this instance, our `button` class is simple (rather unrealistically
 simple in fact, but nevermind that):
```
Comment (christinerose) [inline data/tutorials/lg_07_objects.md:403]: If it's unrealistically simple, does it benefit the reader? If so, does mentioning it's unrealistically simple downplay its importance?

File: data/tutorials/lg_07_objects.md [hunk 32]
```diff
@@ -434,24 +440,24 @@ class button :
...
- `callback` variable is either `None` or `Some f`, in other words it
- has type `(unit -> unit) option`. Reread the previous chapter if you
+ `callback` variable is either `None` or `Some f`, meaning it
+ has type `(unit -> unit) option`. Reread the [previous chapter](https://ocaml.org/docs/functors) if you
  are unsure about this.
 1. Notice a strange thing about the `callback` variable. It's defined
  as an argument to the class, but any method can see and use it. In
  other words, the variable is supplied when the object is
- constructed, but persists over the lifetime of the object.
+ constructed, and it also persists over the lifetime of the object.
```
Comment (christinerose) [inline data/tutorials/lg_07_objects.md:458]:
```suggestion
 constructed, and it also persists over the object's lifetime.
```

File: data/tutorials/lg_07_objects.md [hunk 33]
```diff
@@ -527,21 +534,19 @@ val b : button = <obj>
...
 Error: This expression has type label but an expression was expected of type
          button
        The first object type has no method add
 ```
-I created a button `b` and a label `l` and I tried to create a list
-containing both, but I got an error. Yet `b` and `l` are both `widget`s,
-so why can't I put them into the same list? Perhaps OCaml can't guess
-that I want a `widget list`? Let's try telling it:
+We created a button `b` and a label `l` and then tried to create a list
+containing both, but we got an error. Yet `b` and `l` are both `widget`s,
```
Comment (christinerose) [inline data/tutorials/lg_07_objects.md:542]:
```suggestion
containing both, but it threw an error. Yet `b` and `l` are both `widget`s,
```

**Comment by cuihtlauac** [comment]
I've updated https://github.com/ocaml/ocaml.org/pull/791 with what seemed unique to this PR. Changes had taken place in parallel there. Reconciliation was needed. I hope I haven't lost any of your suggestions in the process.

Thanks again for your feedback, and apologizes for the unpleasant review experience. 

staging is up to date w.r.t. #791

File: data/tutorials/lg_07_objects.md [hunk 34]
```diff
@@ -589,19 +594,19 @@ classes).
...
 
 ### Immediate Objects and Object Types
-Objects can be used instead of records, and have some nice properties
+Objects can be used instead of records. Plus, they have some nice properties
 that can make them preferable to records in some cases. We saw that the
-canonical way of creating objects is to first define a class, and use
+canonical way of creating objects is to first define a class, then use
 this class to create individual objects. This can be cumbersome in some
-situations since class definitions are more than a type definition and
+situations becaus class definitions are more than a type definition and
```
Comment (christinerose) [inline data/tutorials/lg_07_objects.md:609]:
```suggestion
situations because class definitions are more than a type definition and
```

**Comment by c-cube** [comment]
I don't think you lost anything! If anything you attributed too many things to me in the last commit :sweat_smile: . I'm closing this.

No worries for the reviewing experience, it's never easy to make this kind of things smooth. Cheers!

File: src/ocamlorg_frontend/pages/install.eml [hunk 35]
```diff
@@ -0,0 +1,173 @@
...
+      class="grid gap-6 xl:grid-cols-2 xl:gap-16"
+      x-show="(operating_system == 'linux_mac_bsd')">
+      <div class="prose">
+        <h2>
+          Install OCaml on Linux, macOS, or *BSD
+        </h2>
+
+        <ol>
+          <li>
+            <h3>Install the <code>opam</code> package manager</h3>
```
Comment (christinerose) [inline src/ocamlorg_frontend/pages/install.eml:29]:
```suggestion
            <h3>Install the opam Package Manager</h3>
```

File: src/ocamlorg_frontend/pages/install.eml [hunk 36]
```diff
@@ -0,0 +1,173 @@
...
+
+            <p>
+            <a target="_blank" href="https://opam.ocaml.org">OCaml's package manager <tt>opam</tt></a> installs both the compiler, as well as any additional packages.
```
Comment (christinerose) [inline src/ocamlorg_frontend/pages/install.eml:32]:
```suggestion
            <a target="_blank" href="https://opam.ocaml.org">OCaml's package manager opam</a> installs both the compiler, as well as any additional packages.
```

File: src/ocamlorg_frontend/pages/install.eml [hunk 37]
```diff
@@ -0,0 +1,173 @@
...
+            Ensure <code>gcc</code>, <code>build-essential</code>, <code>curl</code>, <code>unzip</code>, and <code>bubblewrap</code> are installed on your system, then
+            run the following in your terminal to download and install the newest version of <code>opam</code>:
```
Comment (christinerose) [inline src/ocamlorg_frontend/pages/install.eml:34]:
```suggestion
            run the following in your terminal to download and install the newest version of opam:
```

File: src/ocamlorg_frontend/pages/install.eml [hunk 38]
```diff
@@ -0,0 +1,173 @@
...
+
+            <%s! Copy_to_clipboard_snippet.render ~id:"install-opam" "bash -c \"sh <(curl -fsSL https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh)\"" %>
+            <a target="_blank" href="https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh" class="flex justify-end gap-1 pt-1 items-center"><%s! Icons.github "h-6 w-6" %> view script on GitHub</a>
+
+            <details>
+              <summary class="flex items-center gap-2 font-semibold">
+                Other installation methods
+                <%s! Icons.chevron_right "h-4 w-4" %>
+              </summary>
+              You can also <a href="<%s Url.tutorial "up-and-running" %>#1-install-opam">install <tt>opam</tt> using your operating system's package manager</a>.
```
Comment (christinerose) [inline src/ocamlorg_frontend/pages/install.eml:45]:
```suggestion
              You can also <a href="<%s Url.tutorial "up-and-running" %>#1-install-opam">install opam using your operating system's package manager</a>.
```

File: src/ocamlorg_frontend/pages/install.eml [hunk 39]
```diff
@@ -0,0 +1,173 @@
...
+
+              If you need the most recent version of <code>opam</code>,
+              you can build <code>opam</code> from sources, following the instructions at
```
Comment (christinerose) [inline src/ocamlorg_frontend/pages/install.eml:48]:
```suggestion
              you can build opam from sources, following the instructions at
```

File: src/ocamlorg_frontend/pages/install.eml [hunk 40]
```diff
@@ -0,0 +1,173 @@
...
+              <a target="_blank" href="https://github.com/ocaml/opam"><tt>opam</tt>'s GitHub repository</a>.
```
Comment (christinerose) [inline src/ocamlorg_frontend/pages/install.eml:49]:
```suggestion
              <a target="_blank" href="https://github.com/ocaml/opam">opam's GitHub repository</a>.
```

File: src/ocamlorg_frontend/pages/install.eml [hunk 41]
```diff
@@ -0,0 +1,173 @@
...
+            </details>
+          </li>
+
+          <li>
+            <h3>Initialise <code>opam</code></h3>
```
Comment (christinerose) [inline src/ocamlorg_frontend/pages/install.eml:54]:
```suggestion
            <h3>Initialise opam</h3>
```

File: src/ocamlorg_frontend/pages/install.eml [hunk 42]
```diff
@@ -0,0 +1,173 @@
...
+
+            <p>
+              <code>opam</code> needs to be initialised, which will create a default
+              <a href="<%s Url.tutorial "up-and-running" %>#3-create-an-opam-switch"><tt>opam</tt> switch</a>.
+              An <code>opam</code> switch is an isolated environment for the OCaml compiler
+              and any packages you install.
+            </p>
```
Comment (christinerose) [inline src/ocamlorg_frontend/pages/install.eml:61]:
```suggestion
              opam needs to be initialised, which will create a default
              <a href="<%s Url.tutorial "up-and-running" %>#3-create-an-opam-switch">opam switch</a>.
              An opam switch is an isolated environment for the OCaml compiler
              and any packages you install.
            </p>
```

File: src/ocamlorg_frontend/pages/install.eml [hunk 43]
```diff
@@ -0,0 +1,173 @@
...
+
+            <%s! Copy_to_clipboard_snippet.render ~id:"init-opam" "opam init" %>
+
+            <p>
+            During <code>opam init</code>, you will be asked if you want to add a hook
+            to your shell to put the tools available in the current <code>opam</code> switch
+            on your PATH.
+            If you answer no, you will need to activate the <code>opam</code> switch
```
Comment (christinerose) [inline src/ocamlorg_frontend/pages/install.eml:69]:
```suggestion
            to your shell to put the tools available in the current opam switch
            on your PATH.
            If you answer no, you will need to activate the opam switch
```

File: src/ocamlorg_frontend/pages/install.eml [hunk 44]
```diff
@@ -0,0 +1,173 @@
...
+
+        <p>Congratulations, you have now installed OCaml!</p>
+      </div>
+      <div>
+        <div class="prose">
+          <h2>Set Up an OCaml Development Environment</h2>
+
+          <p>
+            The <a href="<%s Url.tutorial "up-and-running" %>#installing-the-ocaml-platform-tools">OCaml Platform Tools</a>,
+            which include the <a target="_blank" href="https://dune.build/">build system <tt>dune</tt></a>,
```
Comment (christinerose) [inline src/ocamlorg_frontend/pages/install.eml:83]:
```suggestion
            which include the <a target="_blank" href="https://dune.build/">build system Dune</a>,
```

File: src/ocamlorg_frontend/pages/install.eml [hunk 45]
```diff
@@ -0,0 +1,173 @@
...
+            complete your OCaml development environment. To install them in your current <code>opam</code> switch, run this command:
```
Comment (christinerose) [inline src/ocamlorg_frontend/pages/install.eml:84]:
```suggestion
            complete your OCaml development environment. To install them in your current opam switch, run this command:
```

File: src/ocamlorg_frontend/pages/install.eml [hunk 46]
```diff
@@ -0,0 +1,173 @@
...
+    <div class="grid gap-6 xl:grid-cols-2 xl:gap-16" x-show="operating_system == 'win'">
+      <div class="prose">
+        <h2>
+          Install OCaml on Windows
+        </h2>
+        <ol>
+          <li>
+            <h3>Use the Diskuv OCaml ("DKML") Windows installer</h3>
+              <p>
+                The installer sets up OCaml 4.14.0 and <a target="_blank" href="https://github.com/ocaml/opam">OCaml's package manager <tt>opam</tt></a>,
```
Comment (christinerose) [inline src/ocamlorg_frontend/pages/install.eml:110]:
```suggestion
                The installer sets up OCaml 4.14.0 and <a target="_blank" href="https://github.com/ocaml/opam">OCaml's package manager opam</a>,
```

File: src/ocamlorg_frontend/pages/install.eml [hunk 47]
```diff
@@ -0,0 +1,173 @@
...
+                If you need a different OCaml version, check out the alternative installation instructions
+                provided in the <a href="<%s Url.ocaml_on_windows %>">"OCaml on Windows" guide</a>.
+              </p>
+
+              <a class="btn !text-white hover:no-underline gap-2" href="https://github.com/diskuv/dkml-installer-ocaml/releases/download/v1.2.0/setup-diskuv-ocaml-windows_x86_64-1.2.0.exe">
+                <%s! Icons.download "h-6 w-6" %> Download DKML Installer
+              </a>
+
+              <p>
+                Before you run the installer: Make sure your Windows username does <b>not</b> contain a space character (e.g. for <code>C:\Users\Jane Smith</code>, OCaml will not install properly).
```
Comment (christinerose) [inline src/ocamlorg_frontend/pages/install.eml:122]:
```suggestion
                Before you run the installer: Make sure your Windows username does <b>not</b> contain a space character (e.g., for <code>C:\Users\Jane Smith</code>, OCaml will not install properly).
```

File: src/ocamlorg_frontend/pages/install.eml [hunk 48]
```diff
@@ -0,0 +1,173 @@
...
+        <p>
+          Congratulations, you have now installed OCaml as well as the <a href="<%s Url.tutorial "up-and-running" %>#installing-the-ocaml-platform-tools">OCaml Platform Tools</a>!
+        </p>
+      </div>
+      <div>
+        <div class="prose">
+          <h2>You have an OCaml Development Environment</h2>
+
+          <p>
+            The <a href="<%s Url.tutorial "up-and-running" %>#installing-the-ocaml-platform-tools">OCaml Platform Tools</a> include the <a target="_blank" href="https://dune.build/">build system <tt>dune</tt></a>,
```
Comment (christinerose) [inline src/ocamlorg_frontend/pages/install.eml:137]:
```suggestion
            The <a href="<%s Url.tutorial "up-and-running" %>#installing-the-ocaml-platform-tools">OCaml Platform Tools</a> include the <a target="_blank" href="https://dune.build/">build system Dune</a>,
```

**Review by christinerose — CHANGES_REQUESTED** [review]
Mostly minor grammar/formatting fixes. Tightened some syntax and found a few typos.

**Review by christinerose — CHANGES_REQUESTED** [review]
Mostly minor grammar/formatting fixes. Tightened some syntax and found a few typos.

**Pull request closed without merging**